web3c/oasis.js

```diff
@@ -154,9 +143,78 @@ class Oasis {
     this.accounts = accounts;
   }
 
+  _setupSubscribe(web3) {
+    this.subscribe = (subscription, options, callback) => {
+      // Only try to decrypt logs.
+      if (subscription !== 'logs') {
+        return web3.eth.subscribe(subscription, options, callback);
+      }
+
+      let wrappedEmitter = new EventEmitter();
+
+      let decryptSubscriptionLog = async (msg, log) => {
+        try {
+          if (await this.isConfidential(log.address)) {
+            let transform = new ConfidentialSendTransform(null, this.keyManager);
+            await transform.tryDecryptLogs([log], false);
+            wrappedEmitter.emit(msg, log);
+          } else {
+            wrappedEmitter.emit(msg, log);
+          }
+        } catch (err) {
+          wrappedEmitter.emit('error', err);
+        }
+      };
+
+      let wrappedCallback = undefined;
+      if (callback) {
+        wrappedCallback = async (err, log) => {
+          if (err) {
+            return callback(err);
+          }
+          if (await this.isConfidential(log.address)) {
+            let transform = new ConfidentialSendTransform(null, this.keyManager);
+            await transform.tryDecryptLogs([log], false);
+            return callback(null, log);
+          } else {
+            return callback(null, log);
+          }
+        };
+      }
+
+      web3
+        .eth
+        .subscribe(subscription, options, wrappedCallback)
+        .on('data', async (log) => {
```

Comment:
is it useful to have these as async/await, or can the functions directly return the promises of the encapsulated functions?

Reply:
Good point. We don't strictly need the async await here. There's nothing to return here. We just need to execute the function. I personally think the async await makes things a little more clear/future proof, for example, if someone tries to add a line immediately under `await decryptSubscriptionLog` that depends on the promise finishing.

Reply:
`.on('data', decryptSubmissionLog.bind(this, 'data'))`?

Reply:
Does that pass the `log` in?.  Looks clever.

Reply:
This appears to work nicely.

Reply:
Done. Resolving.